### PR TITLE
Gateway customization plugin

### PR DESCRIPTION
**Pull request by vvpyshnenko**
Issue https://github.com/scalecube/scalecube-js/issues/184

### CODE REVIEW

File: packages/rsocket-ws-gateway/src/createGatewayProxy.ts [hunk 1]
```diff
@@ -9,9 +9,24 @@ interface Proxy {
   [state: string]: any;
 }
 
-export function createGatewayProxy(url: string, definition: MicroserviceApi.ServiceDefinition): Promise<Proxy>;
-export function createGatewayProxy(url: string, definition: MicroserviceApi.ServiceDefinition[]): Promise<Proxy[]>;
-export function createGatewayProxy(url: string, definitions: any): any {
+export function createGatewayProxy(
+  url: string,
+  definition: MicroserviceApi.ServiceDefinition,
+  requestResponse?: any,
```
Comment (IdoJS) [inline packages/rsocket-ws-gateway/src/createGatewayProxy.ts]: add API for requestResponse
```suggestion
  requestResponse?: RequestResponse,
```

File: packages/rsocket-ws-gateway/src/createGatewayProxy.ts [hunk 2]
```diff
@@ -9,9 +9,24 @@ interface Proxy {
...
+  requestStream?: any
+): Promise<Proxy>;
+export function createGatewayProxy(
+  url: string,
+  definition: MicroserviceApi.ServiceDefinition[],
+  requestResponse?: any,
+  requestStream?: any
```
Comment (IdoJS) [inline packages/rsocket-ws-gateway/src/createGatewayProxy.ts]: add API for requestStream
```suggestion
  requestStream?: RequestStream
```

**Comment by IdoJS** [comment]
maybe it is better to have the gateway-client in a separate package

File: packages/rsocket-ws-gateway/src/requestResponse.ts [hunk 3]
```diff
@@ -1,9 +1,11 @@
 import { Single } from 'rsocket-flowable';
 import { packError } from './utils';
 
-export const requestResponse = (payload, serviceCall) => {
+export const requestResponse = (payload, serviceCall, customCb) => {
```
Comment (idanilt) [inline packages/rsocket-ws-gateway/src/requestResponse.ts]:
```
const singleHandler = (subscriber) => {...};
export const requestResponse = (payload, serviceCall, handler = singleHandler) => {
 return new Single(handler);
}
```

File: packages/rsocket-ws-gateway/src/api/Gateway.ts [hunk 4]
```diff
@@ -1,7 +1,11 @@
 import { MicroserviceApi } from '@scalecube/api';
 
+export type GatewayPluginType = (serviceCall: MicroserviceApi.ServiceCall, data: any, subscriber: any) => void;
```
Comment (idanilt) [inline packages/rsocket-ws-gateway/src/api/Gateway.ts]: I would change this name `GatewayPluginType` to `RequestHandler`

I don't like `packages/rsocket-ws-gateway/src/createGatewayProxy.ts` there is room to improve

But it's good enough for now

File: packages/rsocket-ws-gateway/src/Gateway.ts [hunk 5]
```diff
@@ -10,10 +10,14 @@ export class Gateway implements GatewayInterface {
...
   private server;
   private transport;
+  private requestResponse;
+  private requestStream;
   constructor(opts: GatewayOptions) {
     const { port } = opts;
     this.port = port || 3000;
     this.transport = new RSocketWebSocketServer({ port: this.port });
+    this.requestResponse = opts.requestResponse;
+    this.requestStream = opts.requestStream;
```
Comment (IdoJS) [inline packages/rsocket-ws-gateway/src/Gateway.ts]: add validation:
this.requestResponse - opts.requestResponse || defaultValue --> check value is function
this.requestStream - same

**Pull request merged**